src/storage/mvcc/reader/point_getter.rs

```diff
@@ -0,0 +1,783 @@
+            // current key. To keep cursor move in forward direction constantly, let's mark this
+            // state. Additionally this protects us from https://github.com/tikv/tikv/issues/3378.
+            self.lock_cursor_valid = false;
+            return Ok(CheckLockResult::NotLocked);
+        }
+        if lock_cursor.key(&mut self.statistics.lock) == user_key.as_encoded().as_slice() {
+            self.statistics.lock.processed += 1;
+            let lock_value = lock_cursor.value(&mut self.statistics.lock);
+            let lock = Lock::parse(lock_value)?;
+            super::util::check_lock(user_key, ts, &lock)
+        } else {
+            Ok(CheckLockResult::NotLocked)
+        }
+    }
+
+    /// Creates the lock cursor if not created. This function will only be called when
+    /// `multi == true`.
+    fn ensure_lock_cursor(&mut self) -> Result<()> {
+        if self.lock_cursor.is_some() {
+            return Ok(());
+        }
+        // Keys will be given in non-descending order, so forward mode cursor is fine.
+        let cursor = CursorBuilder::new(&self.snapshot, CF_LOCK)
+            .fill_cache(self.fill_cache)
+            .build()?;
+        self.lock_cursor = Some(cursor);
+        self.lock_cursor_valid = true;
+        Ok(())
+    }
+
+    /// Creates the default cursor if not created. This function will only be called when
+    /// `multi == true`.
+    fn ensure_default_cursor(&mut self) -> Result<()> {
+        if self.default_cursor.is_some() {
+            return Ok(());
+        }
+        let cursor = CursorBuilder::new(&self.snapshot, CF_DEFAULT)
+            .fill_cache(self.fill_cache)
+            .build()?;
+        self.default_cursor = Some(cursor);
+        Ok(())
+    }
+
+    /// Load the value.
+    ///
+    /// First, a correct version info in the Write CF will be sought. Then, value will be loaded
+    /// from Default CF if necessary.
+    fn load_data(&mut self, user_key: &Key, ts: u64) -> Result<Option<Value>> {
+        if !self.write_cursor_valid {
+            return Ok(None);
+        }
+
+        // Seek to `${user_key}_${ts}`.
+        if !self
+            .write_cursor
+            .near_seek(&user_key.clone().append_ts(ts), &mut self.statistics.write)?
+        {
+            // If we seek to nothing, it means no write `key >= ${user_key}_${ts}`.
+            // - If later we want to get a key >= current key, due to the above conclusion we can
+            //   quit directly.
+            // - If later we want to get a key < current key, we should prohibit this call.
+            //   Returning nothing directly is safer than some undefined behaviour.
+            // So in all scenarios we should not provide results in future calls when we enter this
+            // branch.
+            self.write_cursor_valid = false;
+        }
+
+        loop {
+            if !self.write_cursor.valid()? {
```

Comment:
Is it ok to use `self.write_cursor_valid` here?

Reply:
`write_cursor_valid` is not equal to `write_cursor.valid()`. `write_cursor_valid` is more like, whether or not the write cursor can be trusted.

Reply:
> `write_cursor_valid` is not equal to `write_cursor.valid()`. `write_cursor_valid` is more like, whether or not the write cursor can be trusted.

So if `self.write_cursor_valid` is not valid, `self.write_cursor.valid()` must be false, we don't need to do extra checkings here?

Reply:
no, if write_cursor_valid is not valid, `self.write_cursor.valid()` may be true.

Reply:
no, if write_cursor_valid is not valid, `self.write_cursor.valid()` may be true.

Reply:
You may take a look at line 215 to see what exactly it means.

Reply:
> You may take a look at line 215 to see what exactly it means.

Could we just return after setting it?